compiler/rustc_ty_utils/src/assoc.rs

```diff
@@ -345,13 +343,51 @@ fn impl_associated_item_for_impl_trait_in_trait(
         fn_has_self_parameter: false,
     });
 
+    impl_assoc_ty.param_env(tcx.param_env(impl_fn_def_id));
+
     // Copy impl_defaultness of the containing function.
     impl_assoc_ty.impl_defaultness(tcx.impl_defaultness(impl_fn_def_id));
 
-    // Copy generics_of the trait's associated item.
-    // FIXME: This is not correct, in particular the parent is going to be wrong. So we would need
-    // to copy from trait_assoc_def_id and adjust things.
-    impl_assoc_ty.generics_of(tcx.generics_of(trait_assoc_def_id).clone());
+    // Copy generics_of the trait's associated item but the impl as the parent.
+    impl_assoc_ty.generics_of({
```

Comment:
Do we need to do the same thing but for traits?

Reply:
Also the contents of this block could possibly be moved into a helper.

Reply:
I don't think so, because in the impls we do all this stuff because we start from `opaque_ty_def_id` that lives inside the trait but we need the parent to be the impl. For the trait given that we also start from `opaque_ty_def_id` and the parent is already the trait, everything is ok I'd assume.

Reply:
> For the trait given that we also start from opaque_ty_def_id and the parent is already the trait, everything is ok I'd assume.

The parent of `opaque_ty_def_id` is the method though.

Reply:
Yes, I've explained it wrong, what I meant is ... to compute the generics for this new synthesized associated items in the trait we need the very same stuff that we need for the opaque. That is the function's generics and the trait's generics and that's what delegating to the opaque would do if I'm correct :).

Reply:
I still think this behavior is subtly wrong, but I guess we'll see it in UI tests (I hope).

Reply:
:+1: let's discuss on follow-ups.